pkg/util/strategicpatch/patch.go

```diff
@@ -35,27 +35,100 @@ import (
 // Some of the content of this package was borrowed with minor adaptations from
 // evanphx/json-patch and openshift/origin.
 
-const specialKey = "$patch"
-const specialValue = "delete"
+const (
+	directiveMarker  = "$patch"
+	deleteDirective  = "delete"
+	replaceDirective = "replace"
+	mergeDirective   = "merge"
+)
+
+// IsPreconditionFailed returns true if the provided error indicates
+// a precondition failed.
+func IsPreconditionFailed(err error) bool {
+	_, ok := err.(errPreconditionFailed)
+	return ok
+}
+
+type errPreconditionFailed struct {
+	message string
+}
+
+func newErrPreconditionFailed(target map[string]interface{}) errPreconditionFailed {
+	s := fmt.Sprintf("precondition failed for: %v", target)
+	return errPreconditionFailed{s}
+}
+
+func (err errPreconditionFailed) Error() string {
+	return err.message
+}
+
+type errConflict struct {
+	message string
+}
+
+func newErrConflict(patch, current []byte) errConflict {
+	s := fmt.Sprintf("patch:\n%v\nconflicts with current:\n%v\n", patch, current)
```

Comment:
Either use `%s` in the format string, or convert byte arrays to string (i.e. `string(patch)`). Otherwise you get illegible output that looks like `patch: [123 34 102 111 111 34 58 34 ...`